processing/src/main/java/io/druid/query/groupby/epinephelinae/Grouper.java

```diff
@@ -0,0 +1,182 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.query.groupby.epinephelinae;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+import java.nio.ByteBuffer;
+import java.util.Iterator;
+
+/**
+ * Groupers aggregate metrics from rows that they typically get from a ColumnSelectorFactory, under
+ * grouping keys that some outside driver is passing in. They can also iterate over the grouped
+ * rows after the aggregation is done.
+ *
+ * They work sort of like a map of KeyType to aggregated values, except they don't support
+ * random lookups.
+ *
+ * @param <KeyType> type of the key that will be passed in
+ */
+public interface Grouper<KeyType extends Comparable<KeyType>>
+{
+  /**
+   * Aggregate the current row with the provided key. May or may not be thread-safe.
+   *
+   * @param key     key object
+   * @param keyHash result of {@link Groupers#hash(Object)} on the key
+   *
+   * @return true if the row was aggregated, false if not due to hitting resource limits
+   */
+  boolean aggregate(KeyType key, int keyHash);
```

Comment:
both spilling and buffer grouper implement it; I thought it would be a nice way to make sure "Grouper" typed objects are always composable without having to encode specific types. I can remove it though if it really bugs you/others.